// react-native-android-widget/lib/commonjs/api/register-widget-task-handler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registerWidgetTaskHandler = registerWidgetTaskHandler;
var _reactNative = require("react-native");
var _AndroidWidget = require("../AndroidWidget");
var _buildWidgetTree = require("./build-widget-tree");
const HEADLESS_TASK_KEY = 'RNWidgetBackgroundTask';
/**
 * Register a task handler that will handle widget actions
 *
 * @param handler {@link WidgetTaskHandler}
 */
function registerWidgetTaskHandler(handler) {
  async function taskProvider(_ref) {
    let {
      widgetAction,
      clickAction,
      clickActionData,
      ...widgetInfo
    } = _ref;
    function renderWidget(widgetComponent) {
      if (widgetAction === 'WIDGET_DELETED') return;
      const lightWidget = 'light' in widgetComponent ? (0, _buildWidgetTree.buildWidgetTree)(widgetComponent.light) : (0, _buildWidgetTree.buildWidgetTree)(widgetComponent);
      const darkWidget = 'dark' in widgetComponent ? (0, _buildWidgetTree.buildWidgetTree)(widgetComponent.dark) : null;
      const config = {
        light: lightWidget,
        dark: darkWidget
      };
      _AndroidWidget.AndroidWidget.drawWidgetById(config, widgetInfo.widgetName, widgetInfo.widgetId);
    }
    await handler({
      widgetInfo,
      widgetAction,
      clickAction,
      clickActionData: clickActionData ? JSON.parse(clickActionData) : {},
      renderWidget
    });
  }
  _reactNative.AppRegistry.registerHeadlessTask(HEADLESS_TASK_KEY, () => taskProvider);
}
