// react-native-android-widget/lib/module/api/register-widget-task-handler.js
import { AppRegistry } from 'react-native';
import { AndroidWidget } from '../AndroidWidget';
import { buildWidgetTree } from './build-widget-tree';
const HEADLESS_TASK_KEY = 'RNWidgetBackgroundTask';
/**
 * Register a task handler that will handle widget actions
 *
 * @param handler {@link WidgetTaskHandler}
 */
export function registerWidgetTaskHandler(handler) {
  async function taskProvider(_ref) {
    let {
      widgetAction,
      clickAction,
      clickActionData,
      ...widgetInfo
    } = _ref;
    function renderWidget(widgetComponent) {
      if (widgetAction === 'WIDGET_DELETED') return;
      const lightWidget = 'light' in widgetComponent ? buildWidgetTree(widgetComponent.light) : buildWidgetTree(widgetComponent);
      const darkWidget = 'dark' in widgetComponent ? buildWidgetTree(widgetComponent.dark) : null;
      const config = {
        light: lightWidget,
        dark: darkWidget
      };
      AndroidWidget.drawWidgetById(config, widgetInfo.widgetName, widgetInfo.widgetId);
    }
    await handler({
      widgetInfo,
      widgetAction,
      clickAction,
      clickActionData: clickActionData ? JSON.parse(clickActionData) : {},
      renderWidget
    });
  }
  AppRegistry.registerHeadlessTask(HEADLESS_TASK_KEY, () => taskProvider);
}
