// react-native-android-widget/lib/module/api/WidgetPreview.js
/* eslint-disable react-native/no-inline-styles */
import React, { useEffect, useState } from 'react';
import { ActivityIndicator, FlatList, Image, Platform, StyleSheet, Text, TouchableNativeFeedback, View } from 'react-native';
import { AndroidWidget } from '../AndroidWidget';
import { buildWidgetTree } from './build-widget-tree';
export function WidgetPreview(_ref) {
  let {
    renderWidget,
    onClick = () => {},
    width,
    height,
    showBorder,
    highlightClickableAreas
  } = _ref;
  const isAndroid = Platform.OS === 'android';
  const [preview, setPreview] = useState();
  const [error, setError] = useState(null);
  useEffect(() => {
    async function init() {
      try {
        const data = await AndroidWidget.createPreview(buildWidgetTree(renderWidget({
          width,
          height
        })), width, height);
        setPreview(data);
        setError(null);
      } catch (e) {
        console.error(e);
        setError((e === null || e === void 0 ? void 0 : e.message) ?? 'Error rendering widget');
      }
    }
    if (isAndroid) {
      init();
    }
    return () => setPreview(null);
  }, [isAndroid, renderWidget, width, height]);
  function onPress(props) {
    var _props$clickActionDat;
    switch (props.clickAction) {
      case 'OPEN_APP':
        console.log('This click will open the app');
        break;
      case 'OPEN_URI':
        console.log(`This click will open ${(_props$clickActionDat = props.clickActionData) === null || _props$clickActionDat === void 0 ? void 0 : _props$clickActionDat.uri}`);
        break;
      default:
        onClick({
          clickAction: props.clickAction,
          clickActionData: props.clickActionData
        });
        break;
    }
  }
  if (!isAndroid) {
    return /*#__PURE__*/React.createElement(PlatformNotAndroid, {
      showBorder: showBorder,
      height: height,
      width: width
    });
  }
  if (error) {
    return /*#__PURE__*/React.createElement(ErrorState, {
      showBorder: showBorder,
      height: height,
      width: width
    });
  }
  return /*#__PURE__*/React.createElement(PreviewContainer, {
    height: height,
    width: width,
    showBorder: showBorder
  }, preview ? /*#__PURE__*/React.createElement(Preview, {
    height: height,
    width: width,
    preview: preview,
    onClick: onPress,
    highlightClickableAreas: highlightClickableAreas
  }) : /*#__PURE__*/React.createElement(ActivityIndicator, {
    size: "large"
  }));
}
function PreviewContainer(_ref2) {
  let {
    showBorder,
    height,
    width,
    children
  } = _ref2;
  return /*#__PURE__*/React.createElement(View, {
    style: {
      height: height + 2,
      width: width + 2,
      borderColor: '#0000ff40',
      alignItems: 'center',
      justifyContent: 'center',
      borderWidth: showBorder ? 1 : 0
    }
  }, children);
}
function Preview(_ref3) {
  let {
    height,
    width,
    preview,
    onClick,
    highlightClickableAreas
  } = _ref3;
  return /*#__PURE__*/React.createElement(View, {
    style: {
      height,
      width,
      position: 'relative'
    }
  }, /*#__PURE__*/React.createElement(Image, {
    source: {
      uri: `data:image/png;base64,${preview.base64Image}`
    },
    style: {
      height,
      width
    }
  }), preview.clickableAreas.map((area, index) => /*#__PURE__*/React.createElement(ClickableAreaButton, {
    key: index,
    area: area,
    onClick: onClick,
    highlightClickableAreas: highlightClickableAreas
  })), preview.collectionAreas.map((area, index) => /*#__PURE__*/React.createElement(CollectionAreaList, {
    key: index,
    area: area,
    onClick: onClick,
    highlightClickableAreas: highlightClickableAreas
  })));
}
function Icon(_ref4) {
  let {
    color,
    height,
    width
  } = _ref4;
  const size = Math.floor(Math.min(height, width) / 3);
  return /*#__PURE__*/React.createElement(View, {
    style: {
      width: size,
      height: size,
      borderRadius: size / 2,
      backgroundColor: color,
      alignItems: 'center',
      justifyContent: 'center',
      marginBottom: 16
    }
  }, /*#__PURE__*/React.createElement(Text, {
    style: {
      fontSize: size / 2,
      color: '#fff'
    }
  }, "!"));
}
function ErrorState(_ref5) {
  let {
    showBorder,
    height,
    width
  } = _ref5;
  return /*#__PURE__*/React.createElement(PreviewContainer, {
    height: height,
    width: width,
    showBorder: showBorder
  }, /*#__PURE__*/React.createElement(Icon, {
    height: height,
    width: width,
    color: "#d32f2f"
  }), /*#__PURE__*/React.createElement(Text, null, "Error rendering widget, see logs"));
}
function PlatformNotAndroid(_ref6) {
  let {
    showBorder,
    height,
    width
  } = _ref6;
  return /*#__PURE__*/React.createElement(PreviewContainer, {
    height: height,
    width: width,
    showBorder: showBorder
  }, /*#__PURE__*/React.createElement(Icon, {
    height: height,
    width: width,
    color: "#ff9966"
  }), /*#__PURE__*/React.createElement(Text, null, "WidgetPreview works only on Android"));
}
function ClickableAreaButton(_ref7) {
  let {
    area,
    onClick,
    highlightClickableAreas
  } = _ref7;
  return /*#__PURE__*/React.createElement(TouchableNativeFeedback, {
    onPress: () => onClick(area)
  }, /*#__PURE__*/React.createElement(View, {
    style: {
      position: 'absolute',
      left: area.left,
      top: area.top,
      width: area.width,
      height: area.height,
      borderColor: 'red',
      borderWidth: highlightClickableAreas ? 1 : 0
    }
  }, highlightClickableAreas ? /*#__PURE__*/React.createElement(ClickableAreaBorder, null) : null));
}
function ClickableAreaBorder() {
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(View, {
    style: styles.clickableHighlightTopLeft
  }), /*#__PURE__*/React.createElement(View, {
    style: styles.clickableHighlightTopRight
  }), /*#__PURE__*/React.createElement(View, {
    style: styles.clickableHighlightBottomLeft
  }), /*#__PURE__*/React.createElement(View, {
    style: styles.clickableHighlightBottomRight
  }));
}
function CollectionAreaList(_ref8) {
  let {
    area,
    onClick,
    highlightClickableAreas
  } = _ref8;
  return /*#__PURE__*/React.createElement(FlatList, {
    style: {
      position: 'absolute',
      left: area.left,
      top: area.top,
      width: area.width,
      height: area.height
    },
    data: area.items,
    renderItem: _ref9 => {
      let {
        item
      } = _ref9;
      return /*#__PURE__*/React.createElement(TouchableNativeFeedback, {
        onPress: () => onClick(item)
      }, /*#__PURE__*/React.createElement(View, {
        style: {
          height: item.height,
          width: item.width,
          borderColor: 'red',
          borderWidth: highlightClickableAreas ? 1 : 0,
          position: 'relative'
        }
      }, highlightClickableAreas ? /*#__PURE__*/React.createElement(ClickableAreaBorder, null) : null, /*#__PURE__*/React.createElement(Image, {
        source: {
          uri: `data:image/png;base64,${item.base64Image}`
        },
        style: {
          height: item.height,
          width: item.width
        }
      }), item.clickableAreas.map((cA, index) => /*#__PURE__*/React.createElement(ClickableAreaButton, {
        key: index,
        area: cA,
        onClick: onClick,
        highlightClickableAreas: highlightClickableAreas
      }))));
    }
  });
}
const styles = StyleSheet.create({
  clickableHighlightTopLeft: {
    borderTopWidth: 1,
    borderLeftWidth: 1,
    borderColor: 'blue',
    height: 8,
    width: 8,
    position: 'absolute',
    left: -1,
    top: -1
  },
  clickableHighlightTopRight: {
    borderTopWidth: 1,
    borderRightWidth: 1,
    borderColor: 'blue',
    height: 8,
    width: 8,
    position: 'absolute',
    right: -1,
    top: -1
  },
  clickableHighlightBottomLeft: {
    borderBottomWidth: 1,
    borderLeftWidth: 1,
    borderColor: 'blue',
    height: 8,
    width: 8,
    position: 'absolute',
    left: -1,
    bottom: -1
  },
  clickableHighlightBottomRight: {
    borderBottomWidth: 1,
    borderRightWidth: 1,
    borderColor: 'blue',
    height: 8,
    width: 8,
    position: 'absolute',
    right: -1,
    bottom: -1
  }
});
