// react-native-android-widget/lib/module/widgets/TextWidget.js
import { convertClickAction } from './utils/click-action';
import { convertColor, convertCommonStyle } from './utils/style.utils';
export function TextWidget(_) {
  return null;
}
TextWidget.convertProps = props => {
  var _props$style, _props$style2, _props$style3, _props$style4, _props$style5, _props$style6, _props$style7, _props$style8;
  return {
    ...convertCommonStyle(props.style ?? {}),
    ...convertClickAction(props),
    text: props.text,
    fontSize: ((_props$style = props.style) === null || _props$style === void 0 ? void 0 : _props$style.fontSize) ?? 12,
    allowFontScaling: props.allowFontScaling ?? true,
    ...((_props$style2 = props.style) !== null && _props$style2 !== void 0 && _props$style2.fontFamily ? {
      fontFamily: props.style.fontFamily
    } : {}),
    ...((_props$style3 = props.style) !== null && _props$style3 !== void 0 && _props$style3.fontStyle ? {
      fontStyle: props.style.fontStyle
    } : {}),
    ...((_props$style4 = props.style) !== null && _props$style4 !== void 0 && _props$style4.fontWeight ? {
      fontWeight: props.style.fontWeight
    } : {}),
    ...((_props$style5 = props.style) !== null && _props$style5 !== void 0 && _props$style5.textAlign ? {
      textAlign: props.style.textAlign
    } : {}),
    ...((_props$style6 = props.style) !== null && _props$style6 !== void 0 && _props$style6.letterSpacing ? {
      letterSpacing: props.style.letterSpacing
    } : {}),
    ...((_props$style7 = props.style) !== null && _props$style7 !== void 0 && _props$style7.adjustsFontSizeToFit ? {
      adjustsFontSizeToFit: props.style.adjustsFontSizeToFit
    } : {}),
    ...((_props$style8 = props.style) !== null && _props$style8 !== void 0 && _props$style8.color ? {
      color: convertColor(props.style.color)
    } : {}),
    ...buildTextShadow(props.style ?? {}),
    ...(props.truncate ? {
      truncate: props.truncate
    } : {}),
    ...(props.maxLines ? {
      maxLines: props.maxLines
    } : {})
  };
};
TextWidget.__name__ = 'TextWidget';
function buildTextShadow(style) {
  if ('textShadowColor' in style || 'textShadowRadius' in style || 'textShadowOffset' in style) {
    var _style$textShadowOffs, _style$textShadowOffs2;
    return {
      shadow: {
        radius: style.textShadowRadius ?? 0,
        dx: ((_style$textShadowOffs = style.textShadowOffset) === null || _style$textShadowOffs === void 0 ? void 0 : _style$textShadowOffs.width) ?? 0,
        dy: ((_style$textShadowOffs2 = style.textShadowOffset) === null || _style$textShadowOffs2 === void 0 ? void 0 : _style$textShadowOffs2.height) ?? 0,
        color: convertColor(style.textShadowColor ?? '#fff')
      }
    };
  }
  return {};
}
