// react-native-android-widget/app.plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = withAndroidWidgets;
var config_plugins_1 = require("expo/config-plugins");
var fs = require("fs");
var path = require("path");
var getMainApplicationOrThrow = config_plugins_1.AndroidConfig.Manifest.getMainApplicationOrThrow;
function withAndroidWidgets(config, params) {
    var _a;
    var projectPaths = {
        platformProjectRoot: '',
        projectRoot: '',
    };
    config = (0, config_plugins_1.withAndroidManifest)(config, function (androidManifestConfig) {
        var mainApplication = getMainApplicationOrThrow(androidManifestConfig.modResults);
        withCollectionService(mainApplication);
        projectPaths.platformProjectRoot =
            androidManifestConfig.modRequest.platformProjectRoot;
        projectPaths.projectRoot = androidManifestConfig.modRequest.projectRoot;
        withConfigurableActivity(mainApplication, params);
        params.widgets.forEach(function (widget) {
            withWidgetReceiver(androidManifestConfig, mainApplication, widget);
        });
        return androidManifestConfig;
    });
    config = withWidgetDescriptions(config, params.widgets);
    config = withFonts(config, projectPaths, (_a = params.fonts) !== null && _a !== void 0 ? _a : []);
    config = withWidgets(config, projectPaths, params.widgets);
    config = withConfigurableActivityClass(config, projectPaths, params.widgets);
    return config;
}
function withCollectionService(mainApplication) {
    var _a, _b;
    mainApplication.service = (_a = mainApplication.service) !== null && _a !== void 0 ? _a : [];
    var alreadyAdded = mainApplication.service.some(function (service) {
        return service.$['android:name'] ===
            'com.reactnativeandroidwidget.RNWidgetCollectionService';
    });
    if (alreadyAdded)
        return;
    (_b = mainApplication.service) === null || _b === void 0 ? void 0 : _b.push({
        $: {
            'android:name': 'com.reactnativeandroidwidget.RNWidgetCollectionService',
            'android:permission': 'android.permission.BIND_REMOTEVIEWS',
        },
    });
}
function withConfigurableActivity(mainApplication, params) {
    var _a, _b;
    var hasConfigurableWidget = params.widgets.some(function (widget) { return !!widget.widgetFeatures; });
    if (hasConfigurableWidget) {
        mainApplication.activity = (_a = mainApplication.activity) !== null && _a !== void 0 ? _a : [];
        var alreadyAdded = mainApplication.activity.some(function (activity) {
            return activity.$['android:name'] === '.WidgetConfigurationActivity';
        });
        if (alreadyAdded)
            return;
        (_b = mainApplication.activity) === null || _b === void 0 ? void 0 : _b.push({
            '$': {
                'android:name': '.WidgetConfigurationActivity',
                'android:exported': 'true',
            },
            'intent-filter': [
                {
                    action: [
                        {
                            $: {
                                'android:name': 'android.appwidget.action.APPWIDGET_CONFIGURE',
                            },
                        },
                    ],
                },
            ],
        });
    }
}
function withConfigurableActivityClass(config, projectPaths, widgets) {
    var hasConfigurableWidget = widgets.some(function (widget) { return !!widget.widgetFeatures; });
    if (hasConfigurableWidget) {
        (0, config_plugins_1.withDangerousMod)(config, [
            'android',
            function (dangerousConfig) {
                var _a, _b, _c;
                var appPackage = path.join(projectPaths.platformProjectRoot, 'android/app/src/main/java/' +
                    ((_b = (_a = config.android) === null || _a === void 0 ? void 0 : _a.package) === null || _b === void 0 ? void 0 : _b.split('.').join('/')));
                var javaFilePath = path.join(appPackage, "/WidgetConfigurationActivity.java");
                var data = "package ".concat((_c = config.android) === null || _c === void 0 ? void 0 : _c.package, ";\n\nimport com.reactnativeandroidwidget.RNWidgetConfigurationActivity;\n\npublic class WidgetConfigurationActivity extends RNWidgetConfigurationActivity {\n}\n");
                fs.writeFileSync(javaFilePath, data);
                return dangerousConfig;
            },
        ]);
    }
    return config;
}
function withWidgetReceiver(config, mainApplication, widget) {
    var _a, _b, _c, _d;
    mainApplication.receiver = (_a = mainApplication.receiver) !== null && _a !== void 0 ? _a : [];
    var receiverName = getReceiverInfo(config, widget).receiverName;
    var alreadyAdded = mainApplication.receiver.some(function (service) { return service.$['android:name'] === receiverName; });
    if (alreadyAdded)
        return;
    (_b = mainApplication.receiver) === null || _b === void 0 ? void 0 : _b.push({
        '$': {
            'android:name': receiverName,
            'android:exported': 'false',
            'android:label': "".concat((_c = widget.label) !== null && _c !== void 0 ? _c : widget.name),
        },
        'intent-filter': [
            {
                action: [
                    {
                        $: {
                            'android:name': 'android.appwidget.action.APPWIDGET_UPDATE',
                        },
                    },
                    {
                        $: {
                            'android:name': "".concat((_d = config.android) === null || _d === void 0 ? void 0 : _d.package, ".WIDGET_CLICK"),
                        },
                    },
                ],
            },
        ],
        'meta-data': {
            $: {
                'android:name': 'android.appwidget.provider',
                'android:resource': "@xml/widgetprovider_".concat(widget.name.toLowerCase()),
            },
        },
    });
}
function withWidgetDescriptions(config, widgets) {
    return (0, config_plugins_1.withStringsXml)(config, function (stringsXml) {
        widgets
            .filter(function (widget) { return !!widget.description; })
            .forEach(function (widget) {
            stringsXml.modResults = config_plugins_1.AndroidConfig.Strings.setStringItem([
                {
                    $: {
                        name: "widget_".concat(widget.name.toLowerCase(), "_description"),
                        translatable: 'false',
                    },
                    _: "".concat(widget.description).replace(/'/g, "\\'"),
                },
            ], stringsXml.modResults);
        });
        return stringsXml;
    });
}
function withFonts(config, androidManifestConfig, fonts) {
    return (0, config_plugins_1.withDangerousMod)(config, [
        'android',
        function (dangerousConfig) {
            if (fonts.length === 0) {
                return dangerousConfig;
            }
            var fontsDir = path.join(androidManifestConfig.platformProjectRoot, 'android/app/src/main/assets/fonts');
            fs.mkdirSync(fontsDir, { recursive: true });
            fonts.forEach(function (font) {
                var fontAssetPath = path.resolve(androidManifestConfig.projectRoot, font);
                var output = path.join(fontsDir, path.basename(fontAssetPath));
                fs.copyFileSync(fontAssetPath, output);
            });
            return dangerousConfig;
        },
    ]);
}
function withWidgets(config, projectPaths, widgets) {
    widgets.forEach(function (widget) {
        withWidget(config, projectPaths, widget);
    });
    return config;
}
function withWidget(config, projectPaths, widget) {
    (0, config_plugins_1.withDangerousMod)(config, [
        'android',
        function (dangerousConfig) {
            withWidgetProviderClass(dangerousConfig, projectPaths, widget);
            withWidgetProviderXml(dangerousConfig, projectPaths, widget);
            withWidgetPreview(projectPaths, widget);
            return dangerousConfig;
        },
    ]);
}
function withWidgetProviderClass(config, projectPaths, widget) {
    var receiverPackage = getReceiverInfo(config, widget).receiverPackage;
    var widgetPackagePath = path.join(projectPaths.platformProjectRoot, 'android/app/src/main/java/' + receiverPackage.split('.').join('/'));
    var javaFilePath = path.join(widgetPackagePath, "/".concat(widget.name, ".java"));
    var data = "package ".concat(receiverPackage, ";\n\nimport com.reactnativeandroidwidget.RNWidgetProvider;\n\npublic class ").concat(widget.name, " extends RNWidgetProvider {\n}\n");
    fs.mkdirSync(widgetPackagePath, { recursive: true });
    fs.writeFileSync(javaFilePath, data);
}
function withWidgetProviderXml(config, projectPaths, widget) {
    var _a, _b;
    var xmlFolderPath = path.join(projectPaths.platformProjectRoot, 'android/app/src/main/res/xml');
    var xmlPath = path.join(xmlFolderPath, "/widgetprovider_".concat(widget.name.toLowerCase(), ".xml"));
    var data = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<appwidget-provider xmlns:android=\"http://schemas.android.com/apk/res/android\"\n    android:minWidth=\"".concat(widget.minWidth, "\"\n    android:minHeight=\"").concat(widget.minHeight, "\"\n").concat(widget.targetCellWidth
        ? "    android:targetCellWidth=\"".concat(widget.targetCellWidth, "\"")
        : '', "\n").concat(widget.targetCellHeight
        ? "    android:targetCellHeight=\"".concat(widget.targetCellHeight, "\"")
        : '', "\n").concat(widget.maxResizeWidth
        ? "    android:maxResizeWidth=\"".concat(widget.maxResizeWidth, "\"")
        : '', "\n").concat(widget.maxResizeHeight
        ? "    android:maxResizeHeight=\"".concat(widget.maxResizeHeight, "\"")
        : '', "\n\n    android:resizeMode=\"").concat((_a = widget.resizeMode) !== null && _a !== void 0 ? _a : 'none', "\"\n\n").concat(widget.description
        ? "    android:description=\"@string/widget_".concat(widget.name.toLowerCase(), "_description\"")
        : '', "\n\n    android:initialLayout=\"@layout/rn_widget\"\n").concat(widget.previewImage
        ? "    android:previewImage=\"@drawable/".concat(widget.name.toLowerCase(), "_preview\"")
        : '', "\n\n").concat(widget.widgetFeatures
        ? "    android:configure=\"".concat((_b = config.android) === null || _b === void 0 ? void 0 : _b.package, ".WidgetConfigurationActivity\"\n    android:widgetFeatures=\"").concat(widget.widgetFeatures, "\"")
        : '', "\n\n    android:updatePeriodMillis=\"").concat(widget.updatePeriodMillis
        ? Math.max(30 * 60 * 1000, widget.updatePeriodMillis)
        : 0, "\"\n    android:widgetCategory=\"home_screen\">\n</appwidget-provider>\n");
    fs.mkdirSync(xmlFolderPath, { recursive: true });
    fs.writeFileSync(xmlPath, data);
}
function withWidgetPreview(projectPaths, widget) {
    if (widget.previewImage) {
        var drawableDir = path.join(projectPaths.platformProjectRoot, 'android/app/src/main/res/drawable');
        fs.mkdirSync(drawableDir, { recursive: true });
        var previewAssetPath = path.resolve(projectPaths.projectRoot, widget.previewImage);
        var output = path.join(drawableDir, "".concat(widget.name.toLowerCase(), "_preview").concat(path.extname(previewAssetPath)));
        fs.copyFileSync(previewAssetPath, output);
    }
}
function getReceiverInfo(config, widget) {
    var _a, _b;
    if (!((_a = config.android) === null || _a === void 0 ? void 0 : _a.package)) {
        throw new Error('Must provide a package for the app');
    }
    if (widget.packageName &&
        !widget.packageName.startsWith(config.android.package)) {
        throw new Error("Package name for widget with name ".concat(widget.name, " must start with ").concat(config.android.package));
    }
    var receiverPackage = (_b = widget.packageName) !== null && _b !== void 0 ? _b : "".concat(config.android.package, ".widget");
    var receiverName = "".concat(receiverPackage.replace(config.android.package, ''), ".").concat(widget.name);
    return { receiverPackage: receiverPackage, receiverName: receiverName };
}
